lib/Dialect/SV/Transforms/HWSpecialize.cpp

```diff
@@ -0,0 +1,298 @@
+//===- HWSpecialize.cpp - hw module specialization ------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+//
+// This transform performs specialization of parametric hw.module's.
+//
+//===----------------------------------------------------------------------===//
+
+#include "PassDetail.h"
+#include "circt/Dialect/Comb/CombOps.h"
+#include "circt/Dialect/HW/HWAttributes.h"
+#include "circt/Dialect/HW/HWOps.h"
+#include "circt/Dialect/SV/SVPasses.h"
+#include "circt/Support/ValueMapper.h"
+#include "mlir/Dialect/SCF/SCF.h"
+#include "mlir/IR/BlockAndValueMapping.h"
+#include "mlir/IR/Builders.h"
+#include "mlir/IR/PatternMatch.h"
+#include "mlir/Transforms/DialectConversion.h"
+#include "llvm/ADT/TypeSwitch.h"
+
+using namespace llvm;
+using namespace mlir;
+using namespace circt;
+using namespace hw;
+
+namespace {
+
+using InstanceParameters = llvm::DenseMap<hw::HWModuleOp, ArrayAttr>;
+
+// Generates a module name by composing the name of 'moduleOp' and the set of
+// provided 'parameters'.
+static std::string generateModuleName(hw::HWModuleOp moduleOp,
+                                      ArrayAttr parameters) {
+  assert(parameters.size() != 0);
+  std::string name = moduleOp.getName().str();
+  for (auto param : parameters) {
+    auto paramAttr = param.cast<ParamDeclAttr>();
+    int64_t paramValue = paramAttr.getValue().cast<IntegerAttr>().getInt();
+    name += "_" + paramAttr.getName().str() + "_" + std::to_string(paramValue);
+  }
+
+  // Uniqueness check
+  auto parentModule = moduleOp->getParentOfType<ModuleOp>();
+  int uniqueCntr = 0;
+  std::string uniqueName = name;
+  while (parentModule.lookupSymbol(uniqueName))
+    uniqueName = name + "_" + std::to_string(uniqueCntr++);
+
+  return uniqueName;
+}
+
+// Returns true if any operand or result of 'op' is parametric.
+static bool isParametricOp(Operation *op) {
+  return llvm::any_of(op->getOperandTypes(), isParametricType) ||
+         llvm::any_of(op->getResultTypes(), isParametricType);
+}
+
+// Narrows 'value' using a comb.extract operation to the width of the
+// hw.array-typed 'array'.
+static Value narrowValueToArrayWidth(OpBuilder &builder, Value array,
```

Comment:
In the case where `value` is defined by a constant, does it make sense to check that the constant is not larger than clog2(arr size)?

Reply:
Since that constant might be parametric, then yes, i'd say it's a worthwhile check - I'll add it!

Reply:
You can't do it (entirely) in the array get verifier since you're bit extracting from the constant, so the array get doesn't see the constant.

Reply:
Hmm, well that should only be the case if the verifier looks at the SSA value itself and doesn't recurse. It would obviously require engineering effort, but i see no reason as to why a verifier cannot be written in the style of e.g. the comb folders, where constant values can be detected through an SSA chain.

Reply:
Ah... But what if the lopping off bits from a large constant and then handing them to an array get is the intended behavior as per an end-designer? It'd be dumb but should be allowed. Actually, should we allow it in the parametric case? Perhaps the only invalid behavior should be if the bits going into array get are statically computed to be larger than the array (so the extract should be respected). Should it be difference in the array get verifier and here? Am I up too late?